### PR TITLE
Allow custom user authentication function

### PR DESCRIPTION
**Pull request by TheBigBadBunce**

### CODE REVIEW

File: exampleApp/src/user/userAuthenticator.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { User } from './user.entity'
+import { TypeOrmModule, getRepositoryToken } from '@nestjs/typeorm'
+import { Repository } from 'typeorm'
+
+export const UserAuthenticator = {
+  imports: [TypeOrmModule.forFeature([User])],
+  // @ts-ignore
```
Comment (williamdclt) [inline exampleApp/src/user/userAuthenticator.ts]: Why is this needed?

File: exampleApp/src/user/userAuthenticator.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { User } from './user.entity'
+import { TypeOrmModule, getRepositoryToken } from '@nestjs/typeorm'
+import { Repository } from 'typeorm'
+
+export const UserAuthenticator = {
```
Comment (williamdclt) [inline exampleApp/src/user/userAuthenticator.ts]: Could we replace all these `Authenticator` with `CredentialValidator`? That's more explicit I think

File: src/adminAuth.module.ts [hunk 3]
```diff
@@ -1,18 +1,42 @@
...
+
+export interface AdminAuthenticator {
+  imports: any[]
+  useFactory: (dep: any) => Authenticator
+  inject: any[]
+}
+
+export const AdminUserAuthenticator = {
+  imports: [TypeOrmModule.forFeature([AdminUser])],
+  // @ts-ignore
```
Comment (williamdclt) [inline src/adminAuth.module.ts]: same

File: src/adminAuth.module.ts [hunk 4]
```diff
@@ -1,18 +1,42 @@
...
+import { Repository } from 'typeorm'
 
 const defaultCoreModule = AdminCoreModuleFactory.createAdminCoreModule({})
 
-export interface AdminAuthModuleConfig {
-  adminCoreModule?: any
-  adminAuthInterface?: AdminAuthenticatorConstructor
+export type Authenticator = (email: string, pass: string) => object | null | Promise<object | null>
+
+export interface AdminAuthenticator {
```
Comment (williamdclt) [inline src/adminAuth.module.ts]:
```suggestion
export interface CredentialValidatorProvider {
```

File: src/adminAuth.module.ts [hunk 5]
```diff
@@ -1,18 +1,42 @@
...
+  useFactory: (dep: any) => Authenticator
+  inject: any[]
+}
+
+export const AdminUserAuthenticator = {
+  imports: [TypeOrmModule.forFeature([AdminUser])],
+  // @ts-ignore
+  useFactory: (adminUserRepository: Repository<AdminUser>) => {
+    return async function validateCredentials(email: string, pass: string) {
+      const user: AdminUser | null = await adminUserRepository.findOne(email)
```
Comment (williamdclt) [inline src/adminAuth.module.ts]: I dont think this works (the password is hashed). You'd want to inject the `AdminUserService` and call its `validate` method

File: src/adminAuth.module.ts [hunk 6]
```diff
@@ -1,18 +1,42 @@
...
+      if (user && pass === user.password) {
+        return user
+      }
+      return null
+    }
+  },
+  inject: [getRepositoryToken(AdminUser)],
+}
+
+interface AdminAuthConfig {
```
Comment (williamdclt) [inline src/adminAuth.module.ts]:
```suggestion
interface AdminAuthModuleConfig {
```

File: src/adminAuth.module.ts [hunk 7]
```diff
@@ -23,17 +47,18 @@ export interface AdminAuthModuleConfig {
...
-      useValue: adminAuthInterface,
+      useFactory: authenticator.useFactory,
+      inject: authenticator.inject,
     }
     return {
       module: AdminAuthModuleFactory,
-      imports: [adminCoreModule],
-      exports: [adminUserServiceProvider, adminAuthInterface],
-      providers: [adminUserServiceProvider, adminAuthInterface],
+      imports: [adminCoreModule, ...authenticator.imports],
```
Comment (williamdclt) [inline src/adminAuth.module.ts]:
```suggestion
      imports: [adminCoreModule, ...(authenticator.imports || [])],
```

File: src/adminAuth.module.ts [hunk 8]
```diff
@@ -1,18 +1,42 @@
...
 
 const defaultCoreModule = AdminCoreModuleFactory.createAdminCoreModule({})
 
-export interface AdminAuthModuleConfig {
-  adminCoreModule?: any
-  adminAuthInterface?: AdminAuthenticatorConstructor
+export type Authenticator = (email: string, pass: string) => object | null | Promise<object | null>
+
+export interface AdminAuthenticator {
+  imports: any[]
```
Comment (williamdclt) [inline src/adminAuth.module.ts]:
```suggestion
  imports?: any[]
```

File: src/adminAuth.module.ts [hunk 9]
```diff
@@ -1,18 +1,42 @@
...
+  useFactory: (dep: any) => Authenticator
+  inject: any[]
```
Comment (williamdclt) [inline src/adminAuth.module.ts]:
```suggestion
  inject?: any[]
```

File: src/adminAuth.module.ts [hunk 10]
```diff
@@ -1,18 +1,37 @@
...
+import AdminUser from './adminUser.entity'
+import { AdminUserService } from './adminUser.service'
 
 const defaultCoreModule = AdminCoreModuleFactory.createAdminCoreModule({})
 
-export interface AdminAuthModuleConfig {
-  adminCoreModule?: any
-  adminAuthInterface?: AdminAuthenticatorConstructor
+export type CredentialValidator = (
+  email: string,
```
Comment (mikerid94) [inline src/adminAuth.module.ts]: email -> username

File: src/tests/login.spec.ts [hunk 11]
```diff
@@ -9,11 +9,13 @@ import { AdminAuthModuleFactory } from '../adminAuth.module'
 describe('login', () => {
   let app: TestApplication
   let document: Document
+  let server: any
```
Comment (mikerid94) [inline src/tests/login.spec.ts]: 'any' :O ?
